crates/sui-indexer-alt/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::BTreeSet, net::SocketAddr, sync::Arc};
+
+use anyhow::{Context, Result};
+use db::{Db, DbConfig};
+use handlers::{pipeline, CommitterConfig, Handler};
+use ingestion::{IngestionConfig, IngestionService};
+use metrics::{IndexerMetrics, MetricsService};
+use models::watermarks::CommitterWatermark;
+use task::graceful_shutdown;
+use tokio::task::JoinHandle;
+use tokio_util::sync::CancellationToken;
+use tracing::info;
+
+pub mod args;
+pub mod db;
+pub mod handlers;
+pub mod ingestion;
+pub mod metrics;
+pub mod models;
+pub mod schema;
+pub mod task;
+
+pub struct Indexer {
+    /// Connection pool to the database.
+    db: Db,
+
+    /// Prometheus Metrics.
+    metrics: Arc<IndexerMetrics>,
+
+    /// Service for serving Prometheis metrics.
+    metrics_service: MetricsService,
+
+    /// Service for downloading and disseminating checkpoint data.
+    ingestion_service: IngestionService,
+
+    /// Parameters for the committers of each pipeline.
+    committer_config: CommitterConfig,
+
+    /// Optional override of the checkpoint lowerbound.
+    first_checkpoint: Option<u64>,
+
+    /// Optional override of the checkpoint upperbound.
+    last_checkpoint: Option<u64>,
+
+    /// Optional override of enabled pipelines.
+    enabled_pipelines: BTreeSet<String>,
+
+    /// Cancellation token shared among all continuous tasks in the service.
+    cancel: CancellationToken,
+
+    /// The checkpoint lowerbound derived from watermarks of pipelines added to the indexer. When
+    /// the indexer runs, it will start from this point, unless this has been overridden by
+    /// [Self::first_checkpoint].
+    first_checkpoint_from_watermark: u64,
```

Comment:
its not immediately clear why we need to have two overrides. is the idea that we start ingesting from the smallest checkpoint, and pipelines sleep until we get to their checkpoint of interest? should this be a single value or a vec corresponding to each pipeline then?

Reply:
This is not a configuration -- it's the field that the `Indexer` uses to accumulate the non-overridden initial checkpoint. It's not convenient to try and store this in one field, because the semantics of the override are that if it is set, then that is the checkpoint you will index from, no matter what. It's tough to get that behaviour with only one field, because when you add a new pipeline, how do you tell from looking at just one number whether that number was an override (in which case, leave it alone), or it's the current first checkpoint given the pipelines subscribed so far (in which case, update it)?

Initially, I implemented this by having a vector of pipeline names, and had a query during the ingestion initialisation to calculate this number, but I went off this idea because pipelines each need to know their initial watermark anyway.

Reply:
ah, i understand now. it's updated when we call `indexer.pipeline()`